### PR TITLE
MDEV-27822 Fix infamous error message

### PR DESCRIPTION
**Pull request by php4fan**
PLEASE

- [x] *The Jira issue number for this PR is: MDEV-_____*

## Description
TODO: fill description here

## How can this PR be tested?

TODO: modify the automated test suite to verify that the PR causes MariaDB to
behave as intended. Consult the documentation on
["Writing good test cases"](https://mariadb.org/get-involved/getting-started-for-developers/writing-good-test-cases-mariadb-server).
In many cases, this will be as simple as modifying one `.test` and one `.result`
file in the `mysql-test/` subdirectory. Without _automated_ tests, future regressions
in the expected behavior can't be automatically detected and verified.

If the changes are not amenable to automated testing, please explain why not and
carefully describe how to test manually.

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

## Backward compatibility
TODO: fill details here, if applicable, or remove the section

### CODE REVIEW

**Comment by grooverdan** [comment]
Why does ER_PARSE_ERROR need to change? I think you are right in that its mostly "starting at", but is it always? Do you expect a better understanding by users in its newer form?

On ER_SYNTAX_ERROR why is removing the description of where to get help important? Sure many people just copy and paste it to forums without reading the message. Is changing the message this going to change that behaviour? Is it going to be recognized in its new form?

**Comment by php4fan** [comment]
**(I highlighted in bold the TL;DR)**

"check the manual that corresponds to your MariaDB server version for the right syntax to use" is an astounding 92 characters to convey exactly 0 information (no it does not give information about where to get help, of course it is in the documentation, where else? That's something I already know hence it's not information).

And what I always found most irritating about this message is that, while wasting those almost 100 characters for saying nothing, it doesn't even tell me enough about where the damn error is.

"Near '<string>'": does it mean I have to look BEFORE or AFTER that string? Actually, first of all I need to look IN that string, which is the only place that was not implied by "near", so it was already misleading to start with. But is it where that string starts or where it ends? When the query is short enough, it may be partly obvious because the string spans to the end of the query, so you kinf of guess the string starts where the error starts, but if I have a very long query and the error is in the middle, I will have no clue, it can be either way (unless of course I've seen the message before with other unrelated errors and learnt how to interpret it). **I distinctly remember having a query like "SELECT foo bar xxx lorem ipsum", getting the error message "Bla bl a bla error near 'xxx'", and looking for half an hour at the "lorem ipsum" part trying to find the error.**

By the way, it should also say "at position N" (or "character N), N being the index of the character where the error was found within the line, because what if the string that it has quoted appears more than once in the query? (unlikely, but possible in a single-line long query). But I don't know how to change that myself so I proposed the easy part of the fix which would already be a significant improvement.

> I think you are right in that its mostly "starting at", but is it always?

I'm pretty sure it is, but that's worth checking indeed.

> Do you expect a better understanding by users in its newer form?

Well yes, that's the whole point.

Every single time I get this error message I'm filled with rage for the amount of time that it makes me waste and how it's capable of conveying so little (and sometimes insufficient) information with so many words. I always stare at it in disbelief that it has been maintained for decades, and the only reason I could think of was because it's so pathetic, so ludicrous, that it's funny. I genuinely can't think of a likely reason except that it's amusing and reminiscent of the history of MySQL (but of course that's not a good reason when it has practical implications).

For a moment I thought you gave me a possible alternative reason worth considering:

> Sure many people just copy and paste it to forums without reading the message

For a moment I thought you had a point there. People are going to copy and paste the error message, and if the error message is changed, they won't find old results. But think about it, what results are they going to get anyway from copying that long message that is exactly the same for any syntax error? Just a ton of completely random results about completely unrelated syntax errors, so, mostly nothing relevant. They'll have to refine their google search anyway.

Either way, is it worth keeping an objectively flawed, inferior error message that contains less (and even arguably misleading) information, just for the sake of maybe finding more results in Google? By that logic, error messages should never be improved (again, now that you make me think about it, that might explain why so many terrible error messages in so many languages take so long to be fixed, or never do; I always assumed the developers didn't gave enough importance to them because they don't make mistakes writing code).


Anyway, I'm sure this needs more work. I seem to understand some test fail (I hope I'm mistaken, do they have the error message hard-coded again and again and compare it to the output??)

Consider it as a bug report. 
Actually I tried reporting the bug in MySQL like 10 years ago, it turned out to have been reported already in 2007, and got no attention whatsoever. It is still open https://bugs.mysql.com/bug.php?id=28466 (although the original report only addresses the redundancy part and not the actual lack of information part).

**Comment by grooverdan** [comment]
> (I highlighted in bold the TL;DR)

Details are good.

> "check the manual that corresponds to your MariaDB server version for the right syntax to use" is an astounding 92 characters 

Fair point, if we tried to put a URL like https://mariadb.com/kb/en/select/ would that be better? I don't know how accurate its possible to make it, but we can add pages there if they are missing.

> "Near ''": does it mean I have to look BEFORE or AFTER that string?

This ties a little into how the parser works, it gets to the point it couldn't continue. So usually its expecting something next, but it also could have started down a different valid syntax and then got to a stopping point.

Starting at "JUNK..." while parsing "SELECT ..." would probably be an improvement. Yes? If we have the starting syntax, the kb url is pretty easy.

> By the way, it should also say "at position N" (or "character N)

That might be reasonably easy.

> completely random results about completely unrelated syntax errors

fair point.

> that might explain why so many terrible error messages in so many languages take so long to be fixed, or never do;

Requests to change longer than "please" are more read. To be fair, searching jira on ER_PARSE_ERROR / ER_SYNTAX_ERROR shows a high number of fixed bugs.

> I always assumed the developers didn't gave enough importance to them because they don't make mistakes writing code

Partially. They don't write hugely complex SQL, and when the error is there for simple cases as you mention its pretty easy, especially if you are writing the function/syntax in question. Also after writing a few syntax changes reading sql/sql_yacc.yy becomes easier. But none of those are good enough reasons not to improve them.

> Consider it as a bug report.

When you aren't filled with rage, do you want to write this up to a bug report on https://jira.mariadb.org?

Are you willing to help further to make a new improved ER_SYNTAX/ER_PARSE error and get the important cases in sql/sql_lex.h sql/sql_yacc.yy sql/sql_lex.cc to use it?

**Comment by php4fan** [comment]
> Fair point, if we tried to put a URL like https://mariadb.com/kb/en/select/ would that be better?

Frankly, I think trying to point to the specific relevant part of documentation is too ambitious. I don't think there's any chance that it can ever be even remotely accurate. It's not even a well defined problem. If I forget to close a parenthesis in a complex nested query with updates and selects and whatever, where should you point me? Nah.

To me the problem is that I need to look for the relevant bit of information hidden within in a long, useless, redundant message. I just want to be able to jump with my eyes right to the point where the information is. If you really think there's value in "Check the manual bla bla bla", just adding a line break after it would accomplish the same that I suggested to accomplish by removing the garbage:

```
You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use starting at:
  "ODRED BY id LIMIT 10"
```
(I would rephrase it because now it doesn't play well with "starting at" but that's not the point here)
To me that long preamble is completely useless, but it doesn't hurt either as long as it doesn't prevent me from quickly identifying the relevant bit.

> This ties a little into how the parser works, it gets to the point it couldn't continue. So usually its expecting something next, but it also could have started down a different valid syntax and then got to a stopping point.

I know that, where the parser blows up is not necessarily where the mistake was made in human terms, but that's not the point. I accept that as a given in programming. Of course the message can be improved in a number of ways (for starters, most parsers tell you "I expected X but found Y instead"), be more specific about what's wrong, give more context etc etc.
But the point of this issue is much simpler: just convey the exact same information that it already contains, but properly convey such information, in a clear and non-misleading way.

The error is saying "The parser blew up exactly here:" but then instead of pointing the finger at where "here" is, it makes some strange gesture with its hands.

**Comment by LinuxJedi** [comment]
Hi @php4fan,

I'm currently going through older pull requests to help with moving them along. For this particular ticket I do think the discourse is important to try an figure out a potentially better way of expressing this error. What I recommend is continuing the discussion in ticket MDEV-27822 and once a plan of action has been agreed a pull request could be made with the change can be opened.

I propose closing this particular PR for now (as the final one would have to be against the current development branch) and opening a new one after the discussion has occurred. How does this sound?

If you have any questions please let me know.

**Comment by LinuxJedi** [comment]
Hi @php4fan,

We have not heard from you in the last month, so I am going to close this for now. But if you are interesting in continuing this work, please feel free to reopen and/or contact me.

**Comment by php4fan** [comment]
Sometimes the error message says 
```
...right syntax to use near ''
```
That's just objectively nonsense.
